embed-code-go/fragmentation/fragmentation.go

```diff
@@ -0,0 +1,211 @@
+//	}
+//
+// ```
+//
+// Fragments with the same name may appear multiple times in the same document.
+//
+// Even if no fragments are defined explicitly, the whole file is always a fragment on its own.
+package fragmentation
+
+import (
+	"bufio"
+	"embed-code/embed-code-go/configuration"
+	"fmt"
+	"os"
+	"path/filepath"
+)
+
+const (
+	FragmentStart = "#docfragment"
+	FragmentEnd   = "#enddocfragment"
+)
+
+// [string] SourcesRoot a full path of the root directory of the source code to be embedded
+// [string] CodeFile a full path of a file to fragment
+type Fragmentation struct {
+	Configuration configuration.Configuration
+	SourcesRoot   string
+	CodeFile      string
+}
+
+//
+// Initializers
+//
+
+func NewFragmentation(
+	codeFileRelative string,
+	config configuration.Configuration,
+) Fragmentation {
+
+	fragmentation := Fragmentation{}
+
+	sourcesRootRelative := config.CodeRoot
+
+	absoluteSourcesRoot, err := filepath.Abs(sourcesRootRelative)
+	fragmentation.SourcesRoot = absoluteSourcesRoot
+	if err != nil {
+		panic(err)
+	}
+
+	absoluteCodeFile, err := filepath.Abs(codeFileRelative)
+	fragmentation.CodeFile = absoluteCodeFile
+	if err != nil {
+		panic(err)
+	}
+
+	fragmentation.Configuration = config
+
+	return fragmentation
+}
+
+//
+// Private methods
+//
+
+func (fragmentation Fragmentation) parseLine(line string, contentToRender []string, fragmentBuilders map[string]*FragmentBuilder) ([]string, map[string]*FragmentBuilder, error) {
+	cursor := len(contentToRender)
+
+	fragmentStarts := GetFragmentStarts(line)
+	fragmentEnds := GetFragmentEnds(line)
+
+	if len(fragmentStarts) > 0 {
+		for _, fragmentName := range fragmentStarts {
+			fragment, exists := fragmentBuilders[fragmentName]
+			if !exists {
+				builder := FragmentBuilder{FileName: fragmentation.CodeFile, Name: fragmentName}
+				fragmentBuilders[fragmentName] = &builder
+				fragment = fragmentBuilders[fragmentName]
+			}
+			fragment.AddStartPosition(cursor)
+		}
+	} else if len(fragmentEnds) > 0 {
+		for _, fragmentName := range fragmentEnds {
+			if fragment, exists := fragmentBuilders[fragmentName]; exists {
+				fragment.AddEndPosition(cursor - 1)
+			} else {
+				return nil, nil, fmt.Errorf("Cannot end the fragment `%s` as it wasn't started.", fragmentName)
+			}
+		}
+	} else {
+		contentToRender = append(contentToRender, line)
+	}
+	return contentToRender, fragmentBuilders, nil
+}
+
+func (fragmentation Fragmentation) targetDirectory() string {
+	fragmentsDir := fragmentation.Configuration.FragmentsDir
+	codeRoot, err := filepath.Abs(fragmentation.Configuration.CodeRoot)
+	if err != nil {
+		panic(fmt.Sprintf("Error calculating absolute path: %v", err))
+	}
+	relativeFile, err := filepath.Rel(codeRoot, fragmentation.CodeFile)
+	if err != nil {
+		panic(fmt.Sprintf("Error calculating relative path: %v", err))
+	}
+	subTree := filepath.Dir(relativeFile)
+	return filepath.Join(fragmentsDir, subTree)
+}
+
+//
+// Public methods
```

Comment:
Throughout this changeset, please make public API methods go first, and private go after.